File: implement-shell-tools/cat/my-cat.py
import sys

args = sys.argv[1:]

flag_n = "-n" in args
flag_b = "-b" in args

files = [arg for arg in args if arg not in ["-n", "-b"]]

line_number = 1

for file in files:
    try:
        with open(file, "r") as f:
            for line in f:
                if flag_b:
                    if line.strip() != "":
                        print(f"{line_number:6}  {line}", end="")
                        line_number += 1
                    else:
                        print(line, end="")
                elif flag_n:
                    print(f"{line_number:6}  {line}", end="")
                    line_number += 1
                else:
                    print(line, end="")
    except FileNotFoundError:
        print(f"cat: {file}: No such file or directory", file=sys.stderr)

File: implement-shell-tools/cat/my-cat.mjs
/* # Implement `cat`

You should already be familiar with the `cat` command line tool.

Your task is to implement your own version of `cat`.

It must act the same as `cat` would, if run from the directory containing this README.md file, for the following command lines:

/* `cat sample-files/1.txt`
/* `cat -n sample-files/1.txt`
/* `cat sample-files/*.txt`
/* `cat -n sample-files/*.txt`
/* `cat -b sample-files/3.txt`

Matching any additional behaviours or flags are optional stretch goals.

We recommend you start off supporting no flags, then add support for `-n`, then add support for `-b`.

 */
import fs from "fs";
import { program } from "commander";

program
  .option("-n")
  .option("-b")
  .argument("<files...>");

program.parse();

const options = program.opts();
const files = program.args;

let lineNumber = 1;

for (const file of files) {
  const content = fs.readFileSync(file, "utf-8");
  const lines = content.split("\n");

  for (const line of lines) {
    if (options.b) {
      if (line !== "") {
        console.log(lineNumber + " " + line);
        lineNumber++;
      } else {
        console.log("");
      }
    } else if (options.n) {
      console.log(lineNumber + " " + line);
      lineNumber++;
    } else {
      console.log(line);
    }
  }
}